### PR TITLE
More seed metrics for costing and scaling.

### PR DESCRIPTION
**Pull request by amshuman-kr**
**What this PR does / why we need it**:
We lack the pod requests, VPA/HVPA recommendations, node utilisation and active shoots metrics in the [aggregate prometheus rules](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml). This makes it hard to evaluate and tune VPA, HVPA and reserve excess capacity optimally.

**Which issue(s) this PR fixes**:
Recording metrics about pod requests, VPA/HVPA recommendations, node utilisation and active shoots in a seed can help to evolve the tuning of auto-scaling of control-planes as well as reserve excess capacity.

**Special notes for your reviewer**:
- I have removed pod-wise aggregate metrics (even the existing ones) to reduce the load on the aggregate prometheus.
- I have introduced a new `kube-state-metrics` in the `garden` namespace to cover all namespaces in the seed cluster.

**Release note**:

```improvement operator
Enhanced monitoring for all seed workloads (shoot control-planes, bootstrap components and extensions) by introducing a new kube-state-metrics in the garden namespace to scrape metrics for all the workloads.
The prometheus in the garden namespace now additionally scrapes the new kube-state-metrics, seed node-exporters, vpa-exporter and hvpa-controller.
New recording rules for the prometheus in the garden namespace to aggregate metrics about usage, request, limits, VPA recommendation, HVPA applied recommendations, node usage, allocatable resources, conditions and counts of pods, nodes and shoots hosted on the seed cluster.
These aggregate metrics are automatically scraped by the aggregate-prometheus.
Some of the pod-wise aggregate metrics have been removed to reduce the load on the aggregate-prometheus.
```

### CODE REVIEW

**Comment by amshuman-kr** [comment]
cc @wyb1

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 1]
```diff
@@ -39,3 +69,48 @@ groups:
...
+    expr: count(kube_pod_info) by (node)
+
+  # Recording rules for allocatable pods per node
+  - record: seed:kube_node_status_allocatable_pods:max
+    expr: max(kube_node_status_allocatable_pods)
+  - record: seed:kube_node_status_allocatable_pods:sum
+    expr: sum(kube_node_status_allocatable_pods)
+
+  # Recording rules for active shoots
+  - record: seed:active_shoots_total:sum
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml:112]: why do we need this?

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 2]
```diff
@@ -3,17 +3,32 @@ groups:
   rules:
   # Recording rules for pod usage
   - record: seed:container_cpu_usage_seconds_total:sum_by_pod
-    expr: sum(rate(container_cpu_usage_seconds_total[5m])) by (pod, namespace)
+    expr: sum(rate(container_cpu_usage_seconds_total[5m])) by (pod, container, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: All container label values are `POD`.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 3]
```diff
@@ -3,17 +3,32 @@ groups:
...
 
   - record: seed:container_memory_working_set_bytes:sum_by_pod
-    expr: sum(container_memory_working_set_bytes) by (pod, namespace)
+    expr: sum(container_memory_working_set_bytes) by (pod, container, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: see above

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 4]
```diff
@@ -3,17 +3,32 @@ groups:
...
 
   - record: seed:container_network_receive_bytes_total:sum_by_pod
-    expr: sum(rate(container_network_receive_bytes_total[5m])) by (pod, namespace)
+    expr: sum(rate(container_network_receive_bytes_total[5m])) by (pod, container, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: see above

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 5]
```diff
@@ -3,17 +3,32 @@ groups:
...
 
   - record: seed:container_network_transmit_bytes_total:sum_by_pod
     expr: sum(rate(container_network_transmit_bytes_total[5m])) by (pod, namespace)
 
+  # Recording rules for pod requests
+  - record: seed:kube_pod_container_resource_requests_cpu_cores:sum_by_pod
+    expr: sum(kube_pod_container_resource_requests_cpu_cores) by (pod, container, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: This comes from the kube-state is each control place. To get this you need to get the federated metrics from each shoot prometheus on the seed.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 6]
```diff
@@ -3,17 +3,32 @@ groups:
...
+
+  - record: seed:kube_pod_container_resource_requests_memory_bytes:sum_by_pod
+    expr: sum(kube_pod_container_resource_requests_memory_bytes) by (pod, container, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: Same as above

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 7]
```diff
@@ -3,17 +3,32 @@ groups:
...
+
+  # Recording rules for vpa recommendations
+  - record: seed:vpa_status_recommendation:sum_by_target
+    expr: sum(vpa_status_recommendation{recommendation="target"}) by (targetName, container, resource, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: This recording rule needs to be part of the aggregate-prometheus recording rules. Then you will only have vpa information of vpa objects in the garden ns. Either the aggregate prometheus needs to federate the data from each prometheus or get all metrics from the vpa exporter.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 8]
```diff
@@ -3,17 +3,32 @@ groups:
...
+
+  # Recording rules for hvpa applied recommendations
+  - record: seed:hvpa_status_applied_vpa_recommendation:sum_by_target
+    expr: sum(hvpa_status_applied_vpa_recommendation{recommendation="target"}) by (targetName, container, resource, namespace)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: same as above

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 9]
```diff
@@ -27,6 +42,21 @@ groups:
   - record: seed:container_network_transmit_bytes_total:sum
     expr: sum(seed:container_network_transmit_bytes_total:sum_by_pod)
 
+  # Recording rules for the sum of the entire seed requests
+  - record: seed:kube_pod_container_resource_requests_cpu_cores:sum
+    expr: sum(seed:kube_pod_container_resource_requests_cpu_cores:sum_by_pod)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: I don't think you can create a recording rule based on another recording rule (unless it has already been federated). Has this changed?

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 10]
```diff
@@ -39,3 +69,48 @@ groups:
...
+  - record: seed:vpa_status_recommendation:sum
+    expr: sum(seed:vpa_status_recommendation:sum_by_target{namespace=~"((shoot-|shoot--)(\\w.+))"})
+
+  # Recording rules for the sum of the entire seed hvpa applied recommendations
+  - record: seed:hvpa_status_applied_vpa_recommendation:sum
+    expr: sum(seed:hvpa_status_applied_vpa_recommendation:sum_by_target{namespace=~"((shoot-|shoot--)(\\w.+))"})
+
+  # Recording rules for node usage
+  - record: seed:node_cpu_seconds_total:sum_by_node
+    expr: sum(rate(node_cpu_seconds_total{mode!="idle"}[5m])) by (node)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: `node_cpu_seconds_total` comes from the node-exporter. Neither this prometheus nor the aggregate prometheus are configured to scrape the node-exporter. Same thing for below metrics

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml [hunk 11]
```diff
@@ -39,3 +69,48 @@ groups:
...
+  - record: seed:node_memory_Active_bytes:sum_by_node
+    expr: sum(node_memory_Active_bytes) by (node)
+
+  # Recording rules for node allocatable
+  - record: seed:kube_node_status_allocatable_cpu_cores:max
+    expr: max(kube_node_status_allocatable_cpu_cores)
```
Comment (wyb1) [inline charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml:94]: metric comes from kube-state-metrics. Neither prometheus is configured to scrape this. Same for below metrics.

**Review by wyb1 — CHANGES_REQUESTED** [review]
These are not the recording rules for the `aggregate-prometheus` but it is the correct place for them (for most of the aggregated rules). You still need to get these metrics into the `aggregate-prometheus` by adding these metrics to the scrape config [here](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/templates/aggregate-prometheus/config.yaml#L61).

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 1 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml:112]: To be able to divide the sum_by_cp metrics by. This can give us avg requests, actuals per actual shoot (since the actual number of active shoots keep fluctuating).

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 2 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: I get the right container names in the shoot prometheus. Is this different in the seed or aggregate prometheus?

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 7 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: 👍 I was testing these and couldn't see these metrics. I will make the changes you suggested and give it another test.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 9 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: I was going by the [above](https://github.com/gardener/gardener/pull/1863/files#diff-1751f67b234069990a5af9a44996a6e0R34) existing rule. But then that was already federated?

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 10 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: This data is available in the shoot prometheus (of the seed in the soil). I was thinking if it could somehow be federated here. If not then I will remove this.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 11 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml:94]: As mentioned [above](https://github.com/gardener/gardener/pull/1863/files#r372048135), it would be nice if this data can be federated here. If not, I will remove this.

**Comment by amshuman-kr** [comment]
@wyb1 Thanks for the review comments and suggestions! I will make the changes.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 2 above)
Comment (wyb1) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml]: Strange, I see the correct container label values now. Maybe I didn't let prometheus run long enough with the new changes.

File: charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml (hunk 1 above)
Comment (wyb1) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml:112]: I don't think we need a recording rule for this. If you do need this you can use your expression or alternatively `count(shoot:availability)`.

Comment (amshuman-kr) [reply charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml:112]: > I don't think we need a recording rule for this.

This is required to analyse the aggregate metrics as the active shoots keeps changing and we can not get per shoot stats without it.

**Comment by amshuman-kr** [comment]
@wyb1 I have completed the changes and tested that the aggregate metrics are showing up. Can you PTAL?

File: charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  namespace: {{ .Release.Namespace }}
+  labels:
+    component: kube-state-metrics
+    type: seed
+
+---
+apiVersion: {{ include "rbacversion" . }}
+kind: ClusterRoleBinding
+metadata:
+  name: kube-state-metrics
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml]: These global resources might already exist. Can we prefix with `gardener.cloud:monitoring` or something like this?

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by rfranzke** [comment]
@amshuman-kr can you please enhance the release note? For me it's very unclear what is happening exactly.

**Comment by amshuman-kr** [comment]
> can you please enhance the release note? For me it's very unclear what is happening exactly.

@rfranzke I have updated the release note. Is it clearer now?

File: charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml (hunk 12 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml]: Done

**Comment by amshuman-kr** [comment]
@rfranzke What's not clear? Should I mention the specific metrics that were added/removed?

**Comment by vpnachev** [comment]
I am the one who is confused 😕. This release note seems too long to read. Could you summarize it?

**Comment by amshuman-kr** [comment]
> I am the one who is confused 😕. This release note seems too long to read. Could you summarize it?

With this PR, prometheus in the garden namespace scapes, usage, requests, limits, vpa recommendations, hvpa applied scalings for all the seed workload (in all the namespaces). Additionally, it also scrapes the seed node-exporters. Based on this data, aggregate metrics have been created for entire seed, only control-planes, per container. Also, there are aggregate node metrics and active shoot metrics to help get utilisation and average per active shoot stats.

Now, that was not shorter 😅

File: charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+kind: ClusterRoleBinding
+metadata:
+  name: gardener.cloud:monitoring:kube-state-metrics
+  labels:
+    component: kube-state-metrics
+    type: seed
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: kube-state-metrics
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml:21]: This should be prefixed as well, I guess.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml (hunk 13 above)
Comment (amshuman-kr) [reply charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml:21]: But the `ClusterRole` [already exists](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/templates/clusterrole-kube-state-metrics.yaml#L8) and is already used in [seed monitoring](https://github.com/gardener/gardener/blob/master/charts/seed-monitoring/charts/core/charts/kube-state-metrics-seed/templates/kube-state-metrics.yaml#L22). Should I still change it?

Comment (rfranzke) [reply charts/seed-bootstrap/charts/kube-state-metrics/templates/kube-state-metrics.yaml:21]: I see.. now, it's okay then. Thanks.

**Review by rfranzke — DISMISSED** [review]

**Comment by amshuman-kr** [comment]
@wyb1 I have made the change to the image metric as discussed in the call and rebased. Can you PTAL?

**Review by wyb1 — APPROVED** [review]
Very nice PR, thanks! /lgtm
